source/slang/slang-ir-redundancy-removal.cpp

```diff
@@ -177,11 +178,29 @@ void removeAvailableInDownstreamModuleDecorations(CodeGenTarget target, IRModule
                     // Gut the function definition, turning it into a declaration
                     for (auto inst : funcInst->getChildren())
                     {
+                        // For HLSL, we can delete the whole block
+                        // but for SPIR-V we have to keep the block with only the
+                        // parameters inside, no other instructions.
                         if (inst->getOp() == kIROp_Block)
                         {
-                            toRemove.add(inst);
+                            if (target == CodeGenTarget::HLSL)
```

Comment:
Why can't HLSL be handled by the same SPIRV path? I would think that HLSL emitter also needs to know the parameters to be able to emit the correct declaration.

Reply:
That hasn't been an issue so far, so it's confusing to me.  HLSL emission manages to preserve the function parameters even when deleting all function blocks.  I don't understand how it's able to.  The code that recognized empty functions and emitted HLSL function declarations existed already and I have yet to see how it worked.

Current code still produces desirable declarations for HLSL:

```
int normalFunc_0(int _0);
int normalFuncUsesGeneric_0(int _0);
```

Reply:
Can we unify the two cases a bit by having spirv emit logic use the func type to emit params too?

We want to have some consistency in our IR in terms of what it means for all backends.

Reply:
It should be doable.  I see emitFuncDecl (c-like) as the example of what HLSL does.
I'm confused why function declarations in SPIR-V require OpFunctionParameters.  It seems redundant to the OpTypeFunction which is already a field of the OpFunction itself.

Reply:
That's probably to allow you to put decorations on the parameters.